### PR TITLE
Process namespace support

### PR DESCRIPTION
**Pull request by m-mohr**
see https://github.com/Open-EO/openeo-api/pull/348

### CODE REVIEW

File: src/process.js [hunk 1]
```diff
@@ -11,8 +11,6 @@ const Utils = require('./utils');
 class BaseProcess {
 
 	constructor(spec) {
-		this.spec = spec; // Keep original specification data
```
Comment (m-mohr) [inline src/process.js:14]: Das hier ist theoretisch breaking, allerdings hab ich die API niemals als öffentlich nutzbare API dokumentiert und somit war die Nutzung IMHO auf eigenes Risiko.

File: src/process.js [hunk 2]
```diff
@@ -86,7 +87,8 @@ class BaseProcess {
 					throw new ProcessGraphError('ProcessGraphParameterMissing', {
 						argument: arg.from_parameter,
 						node_id: node.id,
-						process_id: node.process_id
+						process_id: node.process_id,
+						namespace: node.namespace
```
Comment (christophfriedrich) [inline src/process.js:91]: Why is here no `|| 'n/a'`?

File: src/process.js [hunk 3]
```diff
@@ -164,13 +169,13 @@ class BaseProcess {
 
 	/* istanbul ignore next */
 	async execute(/*node*/) {
-		throw "execute not implemented yet";
+		throw new Error(`execute not implemented yet for process '${this.id}' (namespace: ${this.namespace})`);
```
Comment (christophfriedrich) [inline src/process.js:172]: What if `this.namespace` is null?

File: src/processgraph.js [hunk 4]
```diff
@@ -473,21 +478,23 @@ class ProcessGraph {
...
 	 * @throws {ProcessGraphError} - ProcessUnsupported
 	 */
-	getProcess(id) {
+	getProcess(id, namespace = null) {
 		if (this.processRegistry === null) {
 			return null;
 		}
 		if (id instanceof ProcessGraphNode) {
 			id = id.process_id;
+			namespace = id.namespace;
```
Comment (christophfriedrich) [inline src/processgraph.js:491]: How can a variable named `id` have attributes? That's not very intuitive.

File: src/node.js [hunk 5]
```diff
@@ -36,6 +37,7 @@ class ProcessGraphNode {
 		let args = Utils.mapObjectValues(this.arguments, arg => Utils.isObject(arg) && typeof arg.toJSON === 'function' ? arg.toJSON() : arg);
 		return Object.assign({}, this.source, {
 			process_id: this.process_id,
+			namespace: this.namespace,
```
Comment (christophfriedrich) [inline src/node.js:40]: `|| null` needed?

File: src/process.js (hunk 1 above)
Comment (christophfriedrich) [reply src/process.js:14]: IMO too, so all good :ok_hand:

File: src/process.js [hunk 6]
```diff
@@ -164,13 +169,13 @@ class BaseProcess {
...
 	}
 
 	/* istanbul ignore next */
 	test() {
 		// Run the tests from the examples
-		throw "test not implemented yet";
+		throw new Error(`test not implemented yet for process '${this.id}' (namespace: ${this.namespace})`);
```
Comment (christophfriedrich) [inline src/process.js:178]: Same here

File: src/process.js (hunk 2 above)
Comment (m-mohr) [reply src/process.js:91]: Danke, wohl übersehen/vergessen.

File: src/process.js (hunk 3 above)
Comment (m-mohr) [reply src/process.js:172]: Although this should never occur to actual users and only to devs (who forget to implement the function), I've made it fall back to 'n/a', too.

File: src/processgraph.js (hunk 4 above)
Comment (m-mohr) [reply src/processgraph.js:491]: Indeed, made this more intuitive.

File: src/node.js (hunk 5 above)
Comment (m-mohr) [reply src/node.js:40]: No, undefined is also fine, which will remove the property.

**Review by christophfriedrich — APPROVED** [review]
Looks good now!

**Pull request merged**